### PR TITLE
Spliting of Checkbox question options into its own column for CSV export

### LINKED ISSUE #5020: Check box Questions: Each answer shows in separate column
**Is your feature request related to a problem? Please describe.**
When a check box button question has multi select answers, all answers show up in one field separated by commas in the CSV which is a problem for reporting.

image

**Describe the solution you'd like**
Have multi select answers show each answer in a separate column of the CSV for each possible answer.

**Additional context**

**Done when**
What conditions need to be met before this issue can be closed?

### PR DESCRIPTION
**Pull request by swatkat1**
### Description
All changes needed to split Checkbox question options into its own column for CSV export.

## Release notes

Spliting of Checkbox question options into its own column for CSV export.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)

### Instructions for manual testing

1. Pull this branch to your local machine
2. Run bin/dev-run. Create Checkbox question and program using it.
3. Login as applicant or guest and sub mit applications to the program.
4. Login as Program admin and download the applications as CSV.
5. Each option in the Checkbox question should appear in its own column.

### Issue(s) this completes

Fixes #5020

### CODE REVIEW

File: server/app/services/export/CsvExporter.java [hunk 1]
```diff
@@ -58,11 +71,58 @@ public void exportRecord(
       ReadOnlyApplicantProgramService roApplicantService,
       Optional<Boolean> optionalEligibilityStatus)
       throws IOException {
-    ImmutableMap<Path, String> answerMap =
-        roApplicantService.getSummaryData().stream()
-            .flatMap(data -> data.scalarAnswersInDefaultLocale().entrySet().stream())
-            .collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+    Map<Path, String> answerDataMap = new HashMap<>();
```
Comment (avaleske) [inline server/app/services/export/CsvExporter.java:74]: **suggestion**: I think we can use an `ImmutableMapBuilder` here instead. Then we won't have to convert to the immutable `AnswerMap` at the end, and can leave this named `answerMap` (since it's a map of answers, not of `AnswerData`s.)

File: server/app/services/export/CsvExporter.java [hunk 2]
```diff
@@ -58,11 +71,58 @@ public void exportRecord(
...
+    for (AnswerData answerData : roApplicantService.getSummaryData()) {
+      if (answerData.questionDefinition().getQuestionType().equals(QuestionType.CHECKBOX)) {
+        String questionName = answerData.questionDefinition().getName();
+        // If the question isn't present in the scalar map, it means, its demographic export and
+        // this question was flagged not to be included in demographic export
```
Comment (avaleske) [inline server/app/services/export/CsvExporter.java:79]: **praise** I didn't even think of this edge case!

File: server/test/services/export/CsvExporterTest.java [hunk 3]
```diff
@@ -124,10 +129,22 @@ public void programCsv_noRepeatedEntities() throws Exception {
         testQuestionBank.getSampleQuestionsForAllTypes().get(QuestionType.CHECKBOX);
     MultiSelectQuestion multiSelectApplicantQuestion =
         getApplicantQuestion(checkboxQuestion.getQuestionDefinition()).createMultiSelectQuestion();
-    String multiSelectHeader =
-        CsvExporterService.pathToHeader(multiSelectApplicantQuestion.getSelectionPath());
-    assertThat(records.get(1).get(multiSelectHeader)).isEqualTo("[toaster, pepper_grinder]");
-    // Check link for uploaded file
+
```
Comment (avaleske) [inline server/test/services/export/CsvExporterTest.java:132]: **question**: What would the lift be to also test "not an option in not in program version"? I see it's tested in the browser tests so that could be a fast follow instead of part of this PR.

I've started some refactors of the test logic that could help here, for the JSON test:
https://github.com/civiform/civiform/blob/32bc2499819f2986eb15db46a31bfc7bc4a1cb18/server/test/services/export/JsonExporterTest.java#L268

File: server/test/repository/ExportServiceRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import services.question.QuestionOption;
+import services.question.QuestionService;
+import services.question.types.MultiOptionQuestionDefinition;
+import services.question.types.QuestionDefinition;
+import services.question.types.QuestionDefinitionBuilder;
+import services.question.types.QuestionDefinitionConfig;
+
+public class ExportServiceRepositoryTest extends ResetPostgres {
+  private ExportServiceRepository repo;
+  // private Applicant applicant;
```
Comment (avaleske) [inline server/test/repository/ExportServiceRepositoryTest.java:23]: **nitpick**: remove commented code

File: server/test/repository/ExportServiceRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    versionRepository = instanceOf(VersionRepository.class);
+  }
+
+  @Test
+  public void getMultiSelectedHeaders_NoPreviousQuestionVersion() {
+    QuestionDefinition questionDefinition =
+        testQuestionBank.applicantKitchenTools().getQuestionDefinition();
+    ImmutableList<String> multiSelectHeaders = repo.getMultiSelectedHeaders(questionDefinition);
+    assertThat(multiSelectHeaders.size()).isEqualTo(3);
+    checkList(multiSelectHeaders, 0, "toaster");
```
Comment (avaleske) [inline server/test/repository/ExportServiceRepositoryTest.java:41]: `containsExactly()` could also do what we're doing with checkList() here. It asserts order.

https://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/ListAssert.html#containsExactly(ELEMENT...)

File: server/test/repository/ExportServiceRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        .isInstanceOf(NoSuchElementException.class)
+        .hasMessage("No value present");
+  }
+
+  private void checkList(ImmutableList<String> multiSelectHeaders, int index, String value) {
+    assertThat(multiSelectHeaders.contains(value)).isTrue();
+    assertThat(multiSelectHeaders.get(index)).isEqualTo(value);
+  }
+
+  private Question createMultiSelectQuestion(
```
Comment (avaleske) [inline server/test/repository/ExportServiceRepositoryTest.java:116]: **praise**: Thanks for pulling this out into a helper function!

**suggestion (not blocking)**: structuring this using the Builder pattern would make this easier to extend or customize in the future (eg if we wanted more than 3 options):
https://github.com/civiform/civiform/blob/32bc2499819f2986eb15db46a31bfc7bc4a1cb18/server/test/services/export/AbstractExporterTest.java#L516-L569

File: server/test/repository/ExportServiceRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        new QuestionDefinitionBuilder(questionWeather.getQuestionDefinition())
+            .setQuestionOptions(newOptionList)
+            .build();
+    questionService.update(toUpdate);
+
+    ImmutableList<String> multiSelectHeaders =
+        repo.getMultiSelectedHeaders(questionWeather.getQuestionDefinition());
+    // not draft versions are not part of the header list
+    assertThat(multiSelectHeaders.size()).isEqualTo(3);
+    versionRepository.publishNewSynchronizedVersion();
```
Comment (avaleske) [inline server/test/repository/ExportServiceRepositoryTest.java:68]: **style**: adding aline break before this line will help separate the two "act and assert" blocks.

(Arranging test logic into "arrange", "act", and "assert" code blocks makes them more readable.)

File: server/test/repository/ExportServiceRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+import org.junit.Test;
+import services.LocalizedStrings;
+import services.question.QuestionOption;
+import services.question.QuestionService;
+import services.question.types.MultiOptionQuestionDefinition;
+import services.question.types.QuestionDefinition;
+import services.question.types.QuestionDefinitionBuilder;
+import services.question.types.QuestionDefinitionConfig;
+
+public class ExportServiceRepositoryTest extends ResetPostgres {
```
Comment (avaleske) [inline server/test/repository/ExportServiceRepositoryTest.java:21]: You've covered all the edge cases I can think of here, and several I hadn't though of, except one: Can we also add a test that makes sure removed options are still in the list?

**Review by avaleske — CHANGES_REQUESTED** [review]
Thank you for all your work on this! It's looking really good!

You've clearly thought through a lot of the edge cases here, so I don't have a lot to comment on. The main thing I'm thinking about is:
- Could reading all the applications into memory cause a problem now or down the road? 


Also, before I do a nitpicky style review, can you take a pass at this with the style guide in mind? https://google.github.io/styleguide/javaguide.html

Some things I'm noticing that might need attention:
- Capitalization and grammar in some comments
- Using `@link{foo}` in java docs

File: server/app/repository/ExportServiceRepository.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+   *     checkbox question
+   * @throws RuntimeException when the question is not of type Checkbox
+   * @throws NoSuchElementException when the active version of the question is missing
+   */
+  public ImmutableList<String> getMultiSelectedHeaders(QuestionDefinition questionDefinition) {
+    if (!questionDefinition.getQuestionType().equals(QuestionType.CHECKBOX)) {
+      throw new RuntimeException("The Question Type is not checkbox");
+    }
+    String questionName = questionDefinition.getName();
+    List<Application> applications =
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/repository/ExportServiceRepository.java:53]: Changed logic

File: server/app/repository/ExportServiceRepository.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        database
+            .find(Application.class)
+            .fetch("applicant.applicantData")
+            .where()
+            .in("lifecycle_stage", ImmutableList.of(LifecycleStage.ACTIVE, LifecycleStage.OBSOLETE))
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/repository/ExportServiceRepository.java:58]: Changed logic

File: server/app/repository/ExportServiceRepository.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            .findList();
+    LinkedHashSet<AnswerData> answerSet = new LinkedHashSet<>();
+    for (Application application : applications) {
+      ReadOnlyApplicantProgramService roApplicantService =
+          applicantService
+              .getReadOnlyApplicantProgramService(application)
+              .toCompletableFuture()
+              .join();
+      roApplicantService
+          .getSummaryData()
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/repository/ExportServiceRepository.java:69]: Changed logic

File: server/app/repository/ExportServiceRepository.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+          .forEach(
+              data -> {
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/repository/ExportServiceRepository.java:71]: Changed logic

File: server/app/repository/ExportServiceRepository.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+              if (e.questionDefinition().getQuestionType().equals(QuestionType.CHECKBOX)
+                  && e.questionDefinition().getName().equals(questionName)
+                  && e.isAnswered()) {
+                e
+                    .applicantQuestion()
+                    .createMultiSelectQuestion()
+                    .getSelectedOptionAdminNames()
+                    .get()
+                    .stream()
+                    .forEach(
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/repository/ExportServiceRepository.java:90]: Changed logic

File: server/app/repository/ExportServiceRepository.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    Version activeVersion = versionRepositoryProvider.get().getActiveVersion();
+    MultiOptionQuestionDefinition currentQuestion =
+        (MultiOptionQuestionDefinition)
+            versionRepositoryProvider
+                .get()
+                .getQuestionByNameForVersion(questionName, activeVersion)
+                .get()
+                .getQuestionDefinition();
+    currentQuestion.getOptions().stream()
+        .forEach(
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/repository/ExportServiceRepository.java:108]: Changed logic

File: server/app/services/export/CsvExporter.java (hunk 1 above)
Comment (swatkat1) [reply server/app/services/export/CsvExporter.java:74]: Done

File: server/app/services/export/CsvExporter.java [hunk 15]
```diff
@@ -58,11 +71,58 @@ public void exportRecord(
...
+        List<String> optionHeaders = checkBoxQuestionScalarMap.get(questionName);
+        // the four options for a value are:
+        // in the selected list and in the question definition and question is answered: Selected
+        // not in the selected list and in the question definition and question is answered: Not
+        // Selected
+        // not in the selected list and in the question definition and question is not answered: Not
+        // Answered
+        // not in the selected list and not in the question definition: Not An Option At Program
+        // Version
+        // (This is both “retired” and “not yet an option at the time of this application”)
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/services/export/CsvExporter.java:92]: Done

File: server/app/services/export/CsvExporterService.java [hunk 16]
```diff
@@ -406,24 +437,66 @@ private CsvExportConfig getDemographicsExporterConfig() {
...
+          }
+          List<String> optionHeaders = currentCheckBoxQuestionScalarMap.get(questionName);
+          optionHeaders.stream()
+              .forEach(
+                  option -> {
+                    Path path =
+                        applicantQuestion
+                            .getApplicantQuestion()
+                            .getContextualizedPath()
+                            .join(String.valueOf(option));
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/services/export/CsvExporterService.java:454]: Its not duplicating. The way data is read is very different. In a normal csv export - we look as summary data. In a Demographic data, we look at programDefinitions and create questionDefinition (L439) and try to export it. We also add tags to it. Initially i thought we can combine it, but demographich is export is for all programs while csv is just for one program. They do thing differently.

File: server/test/repository/ExportServiceRepositoryTest.java (hunk 8 above)
Comment (swatkat1) [reply server/test/repository/ExportServiceRepositoryTest.java:21]: Sure thing.

File: server/test/services/export/CsvExporterTest.java (hunk 3 above)
Comment (swatkat1) [reply server/test/services/export/CsvExporterTest.java:132]: Yes may be a fast followup.

File: server/app/repository/ExportServiceRepository.java [hunk 17]
```diff
@@ -1,116 +1,59 @@
...
-    this.versionRepositoryProvider = checkNotNull(versionRepositoryProvider);
-    this.applicantService = checkNotNull(applicantService);
   }
 
   /**
-   * This method queries for all submitted applications in the database ordered by their createTime
-   * and checks if they have used the question passed in the input param. If the question is
-   * present, it creates a unique set of all options ever picked along with the question's active
-   * version options and creates the return list.
+   * This method queries for all unique option admin names ordered by is creation time.
```
Comment (avaleske) [inline server/app/repository/ExportServiceRepository.java:21]:
```suggestion
   * Queries for all unique option admin names ordered by their creation time.
```

https://google.github.io/styleguide/tsguide.html#method-and-function-comments
>Method descriptions begin with a verb phrase that describes what the method does. This phrase is not an imperative sentence, but instead is written in the third person, as if there is an implied This method ... before it.

File: server/app/services/export/CsvExporterService.java [hunk 18]
```diff
@@ -280,26 +286,55 @@ private CsvExportConfig generateDefaultCsvConfig(
...
     }
-    return new CsvExportConfig() {
-      @Override
-      public ImmutableList<Column> columns() {
-        return columnsBuilder.build();
-      }
-    };
+
+    // we cache the checkboxQuestionNameToOptionsMap in csvExportConfig to help us fill the column
+    // values
```
Comment (avaleske) [inline server/app/services/export/CsvExporterService.java:332]: **nit (style)**: 
```suggestion
    // We cache the checkboxQuestionNameToOptionsMap in csvExportConfig to help us fill the column
    // values
```

File: server/app/services/export/enums/MultiOptionSelectionExportType.java [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * the four options for a value are:
+ *
+ * <p>in the selected list and in the question definition and question is answered: SELECTED not in
+ * the selected list and in the question definition and question is answered: NOT_SELECTED not in
+ * the selected list and in the question definition and question is not answered: NOT_ANSWERED not
+ * in the selected list and not in the question definition: NOT_AN_OPTION_AT_PROGRAM_VERSION (or
+ * similar. This is both “retired” and “not yet an option at the time of this application”)
+ */
```
Comment (avaleske) [inline server/app/services/export/enums/MultiOptionSelectionExportType.java:11]: **Nit (style)** Capitalization. Also can we format this as a list? We might need a line break or a `<p>` or a `<li>` so the formatter doesn't mess it up, though.

Alternatively, we could add them as line comments on each line in the enum

File: server/test/repository/ExportServiceRepositoryTest.java (hunk 6 above)
Comment (avaleske) [reply server/test/repository/ExportServiceRepositoryTest.java:152]: Can you file a bug and put it in the TODO, like `TODO(#5555)`?

**Review by avaleske — CHANGES_REQUESTED** [review]
Last few things!

**Review by avaleske — CHANGES_REQUESTED** [review]
Just a few formatting fixes and I noticed a few places we're using `forEach()` instead of `forEachOrdered()`.

Note that `myList.forEach()` is ordered when called on the list directly, but `myList.stream().forEach()` doesn't guarantee order. In that case, we need to call `myList.stream().forEachOrdered()`.

File: server/app/services/export/CsvExporterService.java [hunk 20]
```diff
@@ -280,26 +286,55 @@ private CsvExportConfig generateDefaultCsvConfig(
...
+                option -> {
+                  Path path = answerData.contextualizedPath().join(String.valueOf(option));
+
+                  columnsBuilder.add(
+                      Column.builder()
+                          .setHeader(pathToHeader(path))
+                          .setJsonPath(path)
+                          .setColumnType(ColumnType.APPLICANT_ANSWER)
+                          .build());
+                });
```
[the remark this answers is not in the document]
Comment (swatkat1) [reply server/app/services/export/CsvExporterService.java:318]: I have added foreachOrdered().

File: server/test/services/export/CsvExporterTest.java (hunk 3 above)
Comment (swatkat1) [reply server/test/services/export/CsvExporterTest.java:133]: Added a testcase for it :)

**Review by avaleske — APPROVED** [review]
Thank you for all your work on this!
